compiler/rustc_builtin_macros/src/cfg_match.rs

```diff
@@ -0,0 +1,264 @@
+    }
+    Ok((does_not_have_wildcard, items, items_offsets))
+}
+
+fn parse_arbitrary_arm_block<'a>(
+    items: &mut Vec<P<Item>>,
+    parser: &mut Parser<'a>,
+) -> PResult<'a, ()> {
+    if parser.eat(&token::OpenDelim(token::Delimiter::Brace)) {
+        loop {
+            let item = match parser.parse_item(ForceCollect::No) {
+                Ok(Some(elem)) => elem,
+                _ => break,
+            };
+            items.push(item);
+        }
+        parser.expect(&token::CloseDelim(token::Delimiter::Brace))?;
+    } else {
+        let Ok(Some(item)) = parser.parse_item(ForceCollect::No) else {
+            return parser.unexpected();
+        };
+        if !parser.eat(&token::Comma) {
+            return Err(parser
+                .sess
+                .create_err(errors::CfgMatchMissingComma { span: parser.token.span }));
+        }
+        items.push(item);
+    }
+    Ok(())
+}
+
+fn parse_cfg_arm<'a>(
+    items: &mut Vec<P<Item>>,
+    parser: &mut Parser<'a>,
+) -> PResult<'a, Option<TokenStream>> {
+    if !parser.eat_keyword(sym::cfg) {
+        return Ok(None);
+    }
+    let TokenTree::Delimited(_, delim, tokens) = parser.parse_token_tree() else {
+        return parser.unexpected();
+    };
+    if delim != token::Delimiter::Parenthesis || !parser.eat(&token::FatArrow) {
+        return Err(parser
+            .sess
+            .create_err(errors::CfgMatchBadArm { span: parser.token.span }));
+    }
+    parse_arbitrary_arm_block(items, parser)?;
+    Ok(Some(tokens))
+}
+
+fn parse_wildcard_arm<'a>(items: &mut Vec<P<Item>>, parser: &mut Parser<'a>) -> PResult<'a, ()> {
+    if !parser.eat_keyword(kw::Underscore) || !parser.eat(&token::FatArrow) {
+        return Err(parser
+            .sess
+            .create_err(errors::CfgMatchBadWildcard { span: parser.token.span }));
+    }
+    parse_arbitrary_arm_block(items, parser)?;
+    Ok(())
+}
+
+struct CfgMatchOutput(Vec<P<Item>>);
+
+impl MacResult for CfgMatchOutput {
+    fn make_impl_items(self: Box<Self>) -> Option<SmallVec<[P<AssocItem>; 1]>> {
+        let mut rslt = SmallVec::with_capacity(self.0.len());
+        rslt.extend(self.0.into_iter().filter_map(|item| {
+            let Item { attrs, id, span, vis, ident, kind, tokens } = item.into_inner();
+            let ItemKind::Fn(fun) = kind else {
+                return None;
+            };
+            Some(P(Item { attrs, id, ident, kind: AssocItemKind::Fn(fun), span, tokens, vis }))
+        }));
+        Some(rslt)
+    }
+
+    fn make_items(self: Box<Self>) -> Option<SmallVec<[P<Item>; 1]>> {
+        Some(<_>::from(self.0))
+    }
+
+    fn make_stmts(self: Box<Self>) -> Option<SmallVec<[Stmt; 1]>> {
+        let mut rslt = SmallVec::with_capacity(self.0.len());
+        rslt.extend(self.0.into_iter().map(|item| {
+            let id = item.id;
+            let span = item.span;
+            Stmt { id, kind: StmtKind::Item(item), span }
+        }));
+        Some(rslt)
+    }
+}
```

Comment:
Only items are acceptable as input but the output can be three different things at the current time and I don't if there are more expected variants. This is a best effort.